### PR TITLE
Added too much functionality

### PR DESCRIPTION
**Pull request by jcsduncan**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.12)
+project(modemctl)
+
+set(CMAKE_CXX_STANDARD 11)
+
+find_package(Boost COMPONENTS system filesystem REQUIRED)
+
+if (NOT Boost_FOUND)
```
Comment (CraigCarey) [inline CMakeLists.txt]: Not really needed, the 'REQUIRED' bit in find_package covers this

File: Modem.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#include <utility>
```
Comment (CraigCarey) [inline Modem.h]: Should be inside the include guards

File: Modem.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+//
+// Created by dunx on 19/01/19.
+//
+
+#ifndef LEARNING_MODEM_H
```
Comment (CraigCarey) [inline Modem.h]: I prefer `pragma once` tbh. Bit easier to maintain

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    message(FATAL_ERROR "Could not find boost!")
+endif ()
+
+message(STATUS "Boost_INCLUDE_DIRS: ${Boost_INCLUDE_DIRS}")
+message(STATUS "Boost_LIBRARIES: ${Boost_LIBRARIES}")
+message(STATUS "Boost_VERSION: ${Boost_VERSION}")
+
+add_executable(modemctl main.cpp Serial.cpp Serial.h Modem.cpp Modem.h at_commands.h utils.cpp utils.h)
```
Comment (CraigCarey) [inline CMakeLists.txt:12]: I'd make a modem library and link main to it

File: Modem.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
+//
```
Comment (CraigCarey) [inline Modem.cpp]: Not a fan of these, can use git to see who created or last modified a file. They're usually massively out of date too.

File: Modem.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+class ModemResponse {
+public:
+    ModemResponse(std::vector<std::string> values, bool ok) : values(std::move(values)), command_ok(ok) {};
+    std::vector<std::string> values;
+    bool command_ok;
+};
+
+
+typedef struct modem_info {
```
Comment (CraigCarey) [inline Modem.h]: This and `ModemResponse` seem to serve similar purposes, why is one a class and the other a struct?

File: Modem.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+ModemResponse Modem::get_response() {
+    /*Todo: implement timeout*/
+    std::vector<std::string> responses;
+    std::string line;
+    while (true) {
+        line = rtrim(readline());
+        if (!line.empty()) {
+            if (line == RESP_OK) {
+                return ModemResponse(responses, !(line == RESP_OK));
```
Comment (CraigCarey) [inline Modem.cpp]: Could be `line != RESP_OK`, or just false, since we can see the condition above

File: utils.cpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        return !std::isspace(ch);
+    }));
+}
+
+// trim from end (in place)
+std::string rtrim(std::string &&s) {
+    s.erase(std::find_if(s.rbegin(), s.rend(), [](int ch) {
+        return !std::isspace(ch);
+    }).base(), s.end());
+    return std::move(s);
```
Comment (CraigCarey) [inline utils.cpp]: Don't explicitly return by move, it prevents copy elision. Look up return value optimization.

**Review by CraigCarey — COMMENTED** [review]
Had a quick butcher's, going out for a bit. Let me know if you want me to carry on

File: Modem.h (hunk 6 above)
Comment (jcsduncan) [reply Modem.h]: I'd personally prefer to use classes for everything like this. Fired a struct in to see how you responded to it, gen

File: Modem.cpp (hunk 5 above)
Comment (jcsduncan) [reply Modem.cpp]: True, will disable it in clion.

File: CMakeLists.txt (hunk 4 above)
Comment (jcsduncan) [reply CMakeLists.txt:12]: That's the plan

File: Modem.h (hunk 3 above)
Comment (jcsduncan) [reply Modem.h]: I'll also adopt

File: Modem.h (hunk 2 above)
Comment (jcsduncan) [reply Modem.h]: True

File: CMakeLists.txt (hunk 1 above)
Comment (jcsduncan) [reply CMakeLists.txt]: true

File: Modem.cpp (hunk 7 above)
Comment (jcsduncan) [reply Modem.cpp]: refactored and missed that.

File: utils.cpp (hunk 8 above)
Comment (jcsduncan) [reply utils.cpp]: yep

**Pull request merged**